ONSdigital/census-contact-centre-service
# Cr 1111 put endpoint should cache events

# Motivation and Context
- Previously, all cached cases in firestore had empty caseEvents, now CR-1111 requires that caseEvents are stored in the cachedCases, as retrieved (and filtered) from RM.
- a bug with the PUT (modify) endpoint has been fixed where the createdDateTime was not being updated.
- endpoints returning cachedCase information have been updated to ensure caseEvents are returned when required.

NOTE: there is a minor cucumber change that goes with this for testing empty caseEvents. This needs to be merged at the same time as the cucumber PR.

# What has changed
- as agreed with @philwhiles , filtering of caseEvents has been done early after retrieval from RM. This assists the implementation of this feature.
- As agreed with @philwhiles  , all CaseDTO and cachedCase caseEvents will always be populated with non-null value, even if empty. This assists best practice coding for avoiding nulls for retrieving collections, and corrects previous inconsistent behaviour
- tests have been altered to ensure we check all caseEvents correctly
- some duplication of code in the tests has been centralised for verifying cases.

# How to test?
- unit tests
- cucumber tests
- Postman, in particular:
   - add a  case to the mock-case service
  - ensure it is retrieved correctly by all endpoints (get by ID, UPRN and caseRef) with caseEvents=true, with correct event whitelist filtering
  - do PUT modify operation and check the updated information is stored in firestore
  - ensure the changes as made by PUT modify operation and correctly retried by the get by ID, UPRN and caseRef with caseEvents = true, with correct whitelist filtering.
